lib/teleterm/gateway/interfaces.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+import (
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// Gateway is a interface defines all gateway functions.
+type Gateway interface {
```

Comment:
Could you explain why we need these extra interfaces in this file vs not using them and making `base`, `db` and `kube` public? Isn't the currently implementation breaking this best practice? https://github.com/golang/go/wiki/CodeReviewComments/#interfaces

Reply:
good question.

Here are my reasons:
- Use of gateway scatters across `api`, `clusters`, `daemon`. ideally, the interface should be used in all these places so mocks can be used in testing. If we define the interface in comsuer, we will repeat a few times.
- The first (and most problematic in my opinion) use of gateway is actually in the gateway lib, the `CLICommandProvider`:
```
// CLICommandProvider provides a CLI command for gateways which support CLI clients.                                
type CLICommandProvider interface {                                
    GetCommand(gateway Gateway) (*exec.Cmd, error)                                
}
```
- I don't think `base` should not be exposed. `db` and `kube` can be exposed as public types but need an common interface.
- provide common "casting" method

So this implementation is kinda similar to the usage of `types.Resource`. But while typing this out, not sure if this pattern justifies in this case.

@smallinsky @ravicious what do you think?

Reply:
Ah, I see now, `gateway/interfaces.go` is needed because you cannot "upgrade" `base` to `db` or `kube`. `types.Resource` works like this only because both `Resource` and something like `AppServer` are both interfaces.

---

I'm afraid I don't know enough about Go to have an accurate opinion on this. Defining those interfaces seemed like a smell to me, hence my original comment.

[Marek asked about splitting Gateway](https://github.com/gravitational/teleport/pull/28920#discussion_r1259708226) into GatewayDatabase and GatewayKube, to which you replied that there's no "db flow" and "kube flow" (like there's a separate, say, AppServer flow and KubeServer flow), there's only "gateway flow". I don't know if this is good or not, maybe to adhere to the Go way of things it should be rearranged.

---

While looking for some examples on how to model hierarchies in Go, I found [Type embedding: Golang's fake inheritance](https://www.dolthub.com/blog/2023-02-22-golangs-fake-inheritance/):

> Last week [a complaint from a gopher about his coworkers bringing Java conventions to a Golang codebase](https://www.reddit.com/r/golang/comments/112vfr3/how_to_deal_with_java_developers_polluting_the_go/) made a minor splash on Reddit. Their complaints:
> > (…) All constructor functions return exported interfaces which actually return unexported concrete types. I'm told this is done for encapsulation purposes, otherwise users will not be forced to use the constructor functions. So there is only ever one implementation of an interface, it is implemented by the lower case struct named the same as the exported interface. I really don't like this pattern.

Isn't this what we do here with `Gateway` and `base`?

Ultimately I'd ask someone more knowledgeable in Go about an opinion.

Reply:
I will merge this for now to unblock other changes. But I will keep an on this. I don't think the problem we are solving here is tied to this gateway struct/interface but more about how teleterm lib flows are structured. I will give it more thought and maybe ask around too.